numba/cuda/target.py

```diff
@@ -34,10 +34,19 @@ def load_additional_registries(self):
     def resolve_value_type(self, val):
         # treat dispatcher object as another device function
         if isinstance(val, dispatcher.Dispatcher):
-            if not val._can_compile:
-                raise ValueError('using cpu function on device '
-                                 'but its compilation is disabled')
-            val = jitdevice(val, debug=val.targetoptions.get('debug'))
+            try:
+                # use cached device function
+                val = val.__cudajitdevice
+            except AttributeError:
+                if not val._can_compile:
+                    raise ValueError('using cpu function on device '
+                                     'but its compilation is disabled')
```

Comment:
answer: fixing this will require more substantial refactoring.  Deferring to later.